core/templates/components/graph-services/graph-layout.service.spec.ts

```diff
@@ -0,0 +1,1207 @@
+    ' maximum allowed graph width', () => {
+    let nodeData = {
+      State5: {
+        depth: 2,
+        offset: 1,
+        reachable: true,
+        x0: 0.29750000000000004,
+        y0: 0.54,
+        xLabel: 0.3875,
+        yLabel: 0.6,
+        id: 'State5',
+        label: 'State5',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State4: {
+        depth: 1,
+        offset: 3,
+        reachable: true,
+        x0: 0.7475,
+        y0: 0.33999999999999997,
+        xLabel: 0.8375,
+        yLabel: 0.4,
+        id: 'State4',
+        label: 'State4',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State3: {
+        depth: 1,
+        offset: 2,
+        reachable: true,
+        x0: 0.5225000000000001,
+        y0: 0.33999999999999997,
+        xLabel: 0.6125,
+        yLabel: 0.4,
+        id: 'State3',
+        label: 'State3',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State2: {
+        depth: 1,
+        offset: 1,
+        reachable: true,
+        x0: 0.29750000000000004,
+        y0: 0.33999999999999997,
+        xLabel: 0.3875,
+        yLabel: 0.4,
+        id: 'State2',
+        label: 'State2',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State1: {
+        depth: 1,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.33999999999999997,
+        xLabel: 0.1625,
+        yLabel: 0.4,
+        id: 'State1',
+        label: 'State1',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      Introduction: {
+        depth: 0,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.14,
+        xLabel: 0.1625,
+        yLabel: 0.2,
+        id: 'Introduction',
+        label: 'Introduction',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      End: {
+        depth: 3,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.7400000000000001,
+        xLabel: 0.1625,
+        yLabel: 0.8,
+        id: 'End',
+        label: 'End',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      }
+    };
+    let graphWidth = sgls.getGraphWidth(
+      AppConstants.MAX_NODES_PER_ROW, AppConstants.MAX_NODE_LABEL_LENGTH);
+    let graphHeight = sgls.getGraphHeight(nodeData);
+    let nodeDataWithPositionValueInPixel = sgls.modifyPositionValues(
+      nodeData, graphWidth, graphHeight);
+    let expectedGraphBoundaries = {
+      bottom: 245.8,
+      left: 40.675000000000004,
+      right: 589.325,
+      top: 34.2
+    };
+
+    let expectedWidth = expectedGraphBoundaries.bottom +
```

Comment:
I don't understand this. Why is width = bottom + top? I don't understand why we do the "+" operation and also this seems like the wrong axis.

Reply:
As we are calculating left and right edge as follows:
```javascript
leftEdge = Math.min(
    nodeData[nodeId].x0 - BORDER_PADDING, leftEdge);
    
rightEdge = Math.max(
  nodeData[nodeId].x0 + BORDER_PADDING + nodeData[nodeId].width,
  rightEdge);
```
which translates to 
```
leftEdge = (x0 of leftmost node) - 5
rightEdge = (x0 of rightmost node) + 5 + w

where w = width of a node,
BORDER_PADDING = 5
```
Let, W be width of the graph
If you consider one node at the center, then 
```
x0 = W/2 - w/2
so, 
leftEdge = W/2 - w/2 - 5
rightEdge = W/2 - w/2 + 5 + w = W/2 + w/2 + 5
leftEdge + rightEdge = W
```
The sum would always be W. But this would be always less than 630 (which is the upper bound of graph width).

I have changed the variable `graphWidth` to `graphWidthUpperBound` to make things clear. I hope this explanation was correct.